controller/api/destination/watcher/ip_watcher.go

```diff
@@ -0,0 +1,357 @@
+				iw.log.Errorf("Service cluster IP conflict: %v, %v", objs[0], objs[1])
+			}
+			if len(objs) == 1 {
+				if svc, ok := objs[0].(*corev1.Service); ok {
+					ss.service = ServiceID{
+						Namespace: svc.Namespace,
+						Name:      svc.Name,
+					}
+				}
+			}
+		}
+		objs, err = iw.k8sAPI.Pod().Informer().GetIndexer().ByIndex(podIPIndex, clusterIP)
+		if err != nil {
+			iw.log.Error(err)
+		} else {
+			pods := []*corev1.Pod{}
+			for _, obj := range objs {
+				if pod, ok := obj.(*corev1.Pod); ok {
+					// Skip pods with HostNetwork.
+					if !pod.Spec.HostNetwork {
+						pods = append(pods, pod)
+					}
+				}
+			}
+			if len(pods) > 1 {
+				iw.log.Errorf("Pod IP conflict: %v, %v", objs[0], objs[1])
+			}
+			if len(pods) == 1 {
+				pod := pods[0]
+				ownerKind, ownerName := iw.k8sAPI.GetOwnerKindAndName(pod, true)
+				ss.pod = PodSet{
+					Pods: map[PodID]Address{
+						PodID{
+							Name:      pod.Name,
+							Namespace: pod.Namespace,
+						}: Address{
+							IP:        pod.Status.PodIP,
+							Port:      0, // Will be set by individual subscriptions
+							Pod:       pod,
+							OwnerName: ownerName,
+							OwnerKind: ownerKind,
+						},
+					},
+					Labels: map[string]string{},
+				}
+			}
+		}
+
+		iw.publishers[clusterIP] = ss
+	}
+	return ss
+}
+
+func (iw *IPWatcher) getServiceSubscriptions(clusterIP string) (ss *serviceSubscriptions, ok bool) {
+	iw.RLock()
+	defer iw.RUnlock()
+	ss, ok = iw.publishers[clusterIP]
+	return
+}
+
+////////////////////////
+/// serviceSubscriptions ///
+////////////////////////
+
+func (ss *serviceSubscriptions) updateService(service *corev1.Service) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	id := ServiceID{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	}
+
+	if id != ss.service {
+		for listener, port := range ss.listeners {
+			ss.endpoints.Unsubscribe(ss.service, port, "", listener)
+			listener.NoEndpoints(true) // Clear out previous endpoints.
+			err := ss.endpoints.Subscribe(id, port, "", listener)
+			if err != nil {
+				ss.log.Warnf("failed to subscribe to %s: %s", id, err)
+				listener.NoEndpoints(true) // Clear out previous endpoints.
+				listener.Add(singletonAddress(ss.clusterIP, port))
+			}
+		}
+		ss.service = id
+		ss.pod = PodSet{}
+	}
+}
+
+func (ss *serviceSubscriptions) deleteService() {
+	ss.Lock()
+	defer ss.Unlock()
+
+	for listener, port := range ss.listeners {
+		ss.endpoints.Unsubscribe(ss.service, port, "", listener)
+		listener.NoEndpoints(true) // Clear out previous endpoints.
+		listener.Add(singletonAddress(ss.clusterIP, port))
+
+	}
+	ss.service = ServiceID{}
+}
+
+func (ss *serviceSubscriptions) updatePod(podSet PodSet) {
+	ss.Lock()
+	defer ss.Unlock()
+
+	for listener, port := range ss.listeners {
+		listener.NoEndpoints(true) // Clear out previous endpoints.
+		if len(podSet.Pods) != 0 {
+			podSetWithPort := withPort(podSet, port)
+			listener.Add(podSetWithPort)
+		}
+	}
+	ss.pod = podSet
+	ss.service = ServiceID{}
+}
```

Comment:
It's no longer necessary to send `podSet` as an argument and just use `ss.pod`, and forgo `ss.pod = podSet`. TIOLI: Might even also move this entire logic into `addPod()` itself...